# CCS3440 – Artificial Intelligence Coursework
## SmartCare Hospital AI Dataset — Option C: Disease Risk Classification

**Problem:** Predict a patient's disease risk category (`Low`, `Medium`, `High`) — a multi-class
classification problem — from demographic, clinical, and operational data.

This notebook covers Tasks 02–08 of the coursework specification:
Dataset Understanding, Preprocessing & Feature Engineering, EDA, Model Development,
Model Evaluation, Explainable AI, and the basis for the AI Prototype.


#**Task 2 - Dataset Understanding**


## 2.1 Import library and Load Data set

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', None)
sns.set_style('whitegrid')
RANDOM_STATE = 42

In [ ]:
from google.colab import drive

# Mount Google Drive
drive.mount('/content/drive')

# Read the dataset
df = pd.read_csv('/content/drive/MyDrive/SmartCare/smartcare_ai_dataset_1000.csv')

print("Dataset loaded successfully!")
df.head()

Mounted at /content/drive
Dataset loaded successfully!


,record_id,patient_id,age,gender,blood_group,department,diagnosis,appointment_date,waiting_days,previous_appointments,missed_previous_appointments,appointment_status,admitted,room_type,length_of_stay_days,previous_admissions,systolic_bp,diastolic_bp,blood_sugar_mg_dl,cholesterol_mg_dl,bmi,lab_tests_count,treatments_count,consultation_fee_lkr,room_charge_lkr,lab_charge_lkr,medicine_charge_lkr,total_bill_lkr,payment_status,payment_method,no_show,readmitted_30_days,disease_risk_level
0,1,P10001,53,Male,A-,General Medicine,Migraine,2025-04-10,10,1,0,Completed,0,NaN,0,1,127,75,117,211,26.1,0,3,2000,0,0,11596,13596,Paid,Insurance,0,0,High
1,2,P10002,26,Male,B-,General Medicine,Diabetes,2025-05-15,2,3,1,Completed,0,NaN,0,0,130,73,136,173,32.8,0,1,2000,0,0,3652,5652,Paid,Insurance,0,0,Medium
2,3,P10003,22,Male,B+,Orthopedics,Back Pain,2025-07-09,22,7,1,No-Show,0,NaN,0,1,141,64,90,176,29.4,1,0,2500,0,1200,2562,6262,Unpaid,Insurance,1,0,Medium
3,4,P10004,44,Female,AB-,Cardiology,Asthma,2025-10-16,16,1,0,Completed,0,NaN,0,0,124,82,126,189,24.9,2,1,2000,0,5000,10262,17262,Paid,Online,0,0,Medium
4,5,P10005,51,Female,O+,Neurology,Hypertension,2025-12-18,12,4,0,Scheduled,0,NaN,0,1,119,81,65,195,27.0,2,0,4000,0,6000,10414,20414,Paid,Cash,0,0,Medium


## 2.2 Check Dataset Dimensions


In [ ]:
# Display Dataset Shape
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 1000
Number of columns: 33


## 2.3 Display Dataset Column Names

In [ ]:
# Display Dataset Column
print("Column names in the dataset:")

for column in df.columns:
    print(column)

Column names in the dataset:
record_id
patient_id
age
gender
blood_group
department
diagnosis
appointment_date
waiting_days
previous_appointments
missed_previous_appointments
appointment_status
admitted
room_type
length_of_stay_days
previous_admissions
systolic_bp
diastolic_bp
blood_sugar_mg_dl
cholesterol_mg_dl
bmi
lab_tests_count
treatments_count
consultation_fee_lkr
room_charge_lkr
lab_charge_lkr
medicine_charge_lkr
total_bill_lkr
payment_status
payment_method
no_show
readmitted_30_days
disease_risk_level


## 2.4 Check Dataset Information and Data Types

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 33 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   record_id                     1000 non-null   int64  
 1   patient_id                    1000 non-null   object 
 2   age                           1000 non-null   int64  
 3   gender                        1000 non-null   object 
 4   blood_group                   1000 non-null   object 
 5   department                    1000 non-null   object 
 6   diagnosis                     1000 non-null   object 
 7   appointment_date              1000 non-null   object 
 8   waiting_days                  1000 non-null   int64  
 9   previous_appointments         1000 non-null   int64  
 10  missed_previous_appointments  1000 non-null   int64  
 11  appointment_status            1000 non-null   object 
 12  admitted                      1000 non-null   int64  
 13  room

## 2.5 Display Summary Statistics

In [ ]:
df.describe()

,record_id,age,waiting_days,previous_appointments,missed_previous_appointments,admitted,length_of_stay_days,previous_admissions,systolic_bp,diastolic_bp,blood_sugar_mg_dl,cholesterol_mg_dl,bmi,lab_tests_count,treatments_count,consultation_fee_lkr,room_charge_lkr,lab_charge_lkr,medicine_charge_lkr,total_bill_lkr,no_show,readmitted_30_days
count,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.00000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.00000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000
mean,500.500000,44.744000,21.851000,2.884000,0.553000,0.330000,1.103000,0.85500,128.423000,78.828000,114.832000,204.814000,25.519100,2.292000,1.805000,2820.500000,2412.50000,4869.600000,7971.000000,18073.600000,0.495000,0.253000
std,288.819436,17.851474,13.040892,1.693343,0.740094,0.470448,1.885191,0.95859,15.486895,10.043669,26.017007,36.645577,4.252852,1.679749,1.596723,837.694146,11140.88026,4122.515127,4629.777041,13710.258789,0.500225,0.434948
min,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,85.000000,50.000000,65.000000,100.000000,14.000000,0.000000,0.000000,1500.000000,0.00000,0.000000,506.000000,2954.000000,0.000000,0.000000
25%,250.750000,33.000000,11.000000,2.000000,0.000000,0.000000,0.000000,0.00000,117.000000,72.000000,97.000000,180.000000,22.700000,1.000000,1.000000,2000.000000,0.00000,2400.000000,4316.500000,11018.000000,0.000000,0.000000
50%,500.500000,44.000000,22.000000,3.000000,0.000000,0.000000,0.000000,1.00000,128.000000,79.000000,114.000000,204.000000,25.600000,2.000000,1.000000,3000.000000,0.00000,3600.000000,7680.000000,15068.500000,0.000000,0.000000
75%,750.250000,57.000000,34.000000,4.000000,1.000000,1.000000,2.000000,1.00000,139.000000,86.000000,132.000000,230.000000,28.300000,3.000000,3.000000,3500.000000,0.00000,6300.000000,11032.000000,20870.000000,1.000000,1.000000
max,1000.000000,90.000000,44.000000,10.000000,4.000000,1.000000,9.000000,5.00000,178.000000,111.000000,201.000000,330.000000,38.800000,10.000000,10.000000,4000.000000,150000.00000,27000.000000,20371.000000,179528.000000,1.000000,1.000000


## 2.6 Load and Display the Data Dictionary

In [ ]:
data_dictionary = pd.read_csv(
    '/content/drive/MyDrive/SmartCare/smartcare_ai_dataset_data_dictionary.csv'
)

data_dictionary

,Column,Description
0,record_id,Unique row identifier
1,patient_id,Synthetic patient identifier
2,age,Patient age in years
3,gender,Patient gender
4,blood_group,Patient blood group
5,department,Hospital department
6,diagnosis,Primary diagnosis category
7,appointment_date,Appointment date
8,waiting_days,Number of days between booking and appointment
9,previous_appointments,Number of previous appointments


## 2.7 Identify and Examine the Target Variable

In [ ]:
target_variable = 'disease_risk_level'

print("Selected target variable:", target_variable)
print("\nTarget class distribution:")
print(df[target_variable].value_counts())

print("\nTarget proportion:\n", df['disease_risk_level'].value_counts(normalize=True).round(3))

Selected target variable: disease_risk_level

Target class distribution:
disease_risk_level
Medium    469
High      400
Low       131
Name: count, dtype: int64

Target proportion:
 disease_risk_level
Medium    0.469
High      0.400
Low       0.131
Name: proportion, dtype: float64


**Observation:** The classes are imbalanced (Medium ≈ 47%, High ≈ 40%, Low ≈ 13%).
We account for this later using `class_weight='balanced'` where supported, stratified
train/test splitting, and by reporting macro-averaged (not just accuracy) metrics.

## 2.8 Identify Input Variables

In [ ]:
# Display Input Variable
excluded_columns = [
    'record_id',
    'patient_id',
    'no_show',
    'readmitted_30_days',
    'disease_risk_level'
]

input_variables = [
    column for column in df.columns
    if column not in excluded_columns
]

print("Selected target variable:")
print(target_variable)

print("\nNumber of potential input variables:", len(input_variables))

print("\nPotential input variables:")
for variable in input_variables:
    print(variable)

Selected target variable:
disease_risk_level

Number of potential input variables: 28

Potential input variables:
age
gender
blood_group
department
diagnosis
appointment_date
waiting_days
previous_appointments
missed_previous_appointments
appointment_status
admitted
room_type
length_of_stay_days
previous_admissions
systolic_bp
diastolic_bp
blood_sugar_mg_dl
cholesterol_mg_dl
bmi
lab_tests_count
treatments_count
consultation_fee_lkr
room_charge_lkr
lab_charge_lkr
medicine_charge_lkr
total_bill_lkr
payment_status
payment_method


## 2.9 Identify Numerical and Categorical Variables

In [ ]:
numerical_variables = df[input_variables].select_dtypes(
    include=['int64', 'float64']
).columns.tolist()

categorical_variables = df[input_variables].select_dtypes(
    include=['object']
).columns.tolist()

print("Numerical variables:")
for variable in numerical_variables:
    print(variable)

print("\nCategorical variables:")
for variable in categorical_variables:
    print(variable)

Numerical variables:
age
waiting_days
previous_appointments
missed_previous_appointments
admitted
length_of_stay_days
previous_admissions
systolic_bp
diastolic_bp
blood_sugar_mg_dl
cholesterol_mg_dl
bmi
lab_tests_count
treatments_count
consultation_fee_lkr
room_charge_lkr
lab_charge_lkr
medicine_charge_lkr
total_bill_lkr

Categorical variables:
gender
blood_group
department
diagnosis
appointment_date
appointment_status
room_type
payment_status
payment_method


## 2.10 Check Unique Values in Categorical Variables

In [ ]:
# Check Categorical Variables Values
print("Number of unique values in categorical variables:")

for variable in categorical_variables:
    print(variable, ":", df[variable].nunique())

Number of unique values in categorical variables:
gender : 2
blood_group : 8
department : 7
diagnosis : 10
appointment_date : 339
appointment_status : 4
room_type : 3
payment_status : 3
payment_method : 4


## 2.11 Display Categories in Categorical Variables

In [ ]:
for variable in categorical_variables:
    unique_count = df[variable].nunique()

    if unique_count <= 20:
        print(f"\n{variable}:")
        print(df[variable].dropna().unique())


gender:
['Male' 'Female']

blood_group:
['A-' 'B-' 'B+' 'AB-' 'O+' 'A+' 'AB+' 'O-']

department:
['General Medicine' 'Orthopedics' 'Cardiology' 'Neurology' 'Pediatrics'
 'Laboratory Services' 'Radiology']

diagnosis:
['Migraine' 'Diabetes' 'Back Pain' 'Asthma' 'Hypertension' 'Fracture'
 'Kidney Infection' 'Fever' 'Chest Pain' 'Pneumonia']

appointment_status:
['Completed' 'No-Show' 'Scheduled' 'Cancelled']

room_type:
['Private Room' 'ICU' 'General Ward']

payment_status:
['Paid' 'Unpaid' 'Partially Paid']

payment_method:
['Insurance' 'Online' 'Cash' 'Card']


## 2.12 Data Quality Findings

In [ ]:
print("Duplicate rows:", df.duplicated().sum())
print("\nMissing values per column:")
missing = df.isnull().sum()
print(missing[missing > 0])
print("\nMissing % of room_type:", round(df['room_type'].isnull().mean()*100, 1), "%")

Duplicate rows: 0

Missing values per column:
room_type    906
dtype: int64

Missing % of room_type: 90.6 %


The dataset contains 1,000 records and 33 attributes. It includes numerical, categorical, and date-related data. Most columns contain 1,000 non-null values. However, the room_type column contains only 94 non-null values and 906 missing values. This is mainly because a room type is relevant only when a patient is admitted to the hospital. Therefore, these missing values may be structurally expected rather than random data errors.

The selected target variable, disease_risk_level, contains three classes: Low, Medium, and High. The class distribution is not fully balanced because the Low-risk category has fewer records than the Medium- and High-risk categories. This imbalance should be considered during model training and evaluation.

## 2.13 Dataset Overview Summary

The SmartCare Hospital AI dataset contains 1,000 hospital records and 33 attributes. It covers patient demographics, appointments, admissions, clinical measurements, treatments, billing details, and three AI prediction targets.

For this project, `disease_risk_level` was selected as the target variable. It is a multiclass target containing Low, Medium, and High disease-risk categories. Identifier columns and the two alternative AI targets were excluded from the potential input-variable list.

The dataset is generally suitable for machine-learning development because it contains no duplicate records and most attributes have complete values. The main missing-data issue occurs in the `room_type` attribute. Data types, numerical statistics, categorical values, input variables, and the selected target variable were examined during the dataset-understanding process.